src/sage/data_structures/stream.py

```diff
@@ -381,21 +381,45 @@ def __getitem__(self, n):
 
         if self._is_sparse:
             try:
-                c = self._cache[n]
+                return self._cache[n]
             except KeyError:
                 c = self.get_coefficient(n)
                 self._cache[n] = c
-        else:
-            i = n - self._offset
-            if i >= len(self._cache):
-                a = len(self._cache) + self._offset
+                if not self._true_order and n == self._approximate_order:
+                # self._approximate_order is not in self._cache if
+                # self._true_order is False
+                    if c:
+                        self._true_order = True
+                        self._approximate_order = n
+                    else:
+                        ao = self._approximate_order + 1
+                        while ao in self._cache:
+                            if self._cache[ao]:
+                                self._true_order = True
+                                break
+                            ao += 1
+                        self._approximate_order = ao
+                return c
+
+        # Dense implementation
+        i = n - self._offset
+        if i >= len(self._cache):
+            a = len(self._cache) + self._offset
+            if self._true_order:
                 # It is important to extend by generator:
                 # self._iter might recurse, and thereby extend the
                 # cache itself, too.
                 self._cache.extend(next(self._iter) for _ in range(a, n+1))
-            c = self._cache[i]
+            else:
+                for _ in range(a, n+1):
+                    c = next(self._iter)
+                    self._cache.append(c)
+                    if c:
+                        self._true_order = True
+                    else:
+                        self._approximate_order += 1
```

Comment:
I don't see how to make inline review suggestions that span such changes like this. It would be nice if things worked much more uniformly than they do...

```suggestion
        # Dense implementation
        i = n - self._offset
        if i >= len(self._cache):
            a = len(self._cache) + self._offset
            if not self._true_order:
                for _ in range(a, n+1):
                    c = next(self._iter)
                    if c:
                        self._true_order = True
                        break
                a = len(self._cache) + self._offset
                self._approximate_order = a
            if self._true_order:
                # It is important to extend by generator:
                # self._iter might recurse, and thereby extend the
                # cache itself, too.
                self._cache.extend(next(self._iter) for _ in range(a, n+1))
```
It is slightly more complicated logic, but it reduces some duplication and should be faster. Basically, it works by updating until it finds the true order then proceeds by a direct extension. If it doesn't find the true order but reaches `n`, then it falls through with a large enough cache.

Reply:
I like the idea, however, as coded it does not work for me.  One reason is that we have to put `c` into the cache.  However, this still breaks `_offset`.  I am working on it.